src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const array = [];
+  const stateClone = { ...state };
+
+  for (const key of actions) {
+    switch (key.type) {
+      case 'addProperties':
+        Object.assign(stateClone, key.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const removeKey of key.keysToRemove) {
+          delete stateClone[removeKey];
+        }
+        break;
+
+      case 'clear':
+        for (const kay in stateClone) {
+          delete stateClone[kay];
+        }
+        break;
+    }
+
+    array.push(Object.assign({}, stateClone));
```

Comment:
[CODE STYLE] - It's better to push a copy of the object to the array at the end of the loop cycle, but outside of the switch block. However, this is already done correctly here.